proposals/gc/Overview.md

```diff
@@ -1 +1,696 @@
+The index is checked at execution time.
+A trap occurs if the index is out of bounds.
+
+Arrays are [allocated](#allocation) with `new` instructions that take a size and an initialization value as operands, yielding a reference:
+```
+(func $g
+  (call $g (new_array $vector (i32.const 0) (i64.const 3.14)))
+)
+```
+Arrays are garbage-collected.
+
+The *length* of an array, i.e., the number of elements, can be inquired via the `load_length` instruction:
+```
+(load_length $vector (get_local $v))
+```
+
+
+### Packed Fields
+
+Fields and elements can have a packed *storage type* `i8` or `i16`:
+```
+(type $s (struct (field $a i8) (field $b i16)))
+(type $buf (array i8))
+```
+The order of fields is not observable, so implementations are free to optimize types by reordering fields or adding gaps for alignment.
+
+Packed fields require special load/store instructions:
+```
+(load_field_packed_s $s $a (...))
+(load_field_packed_u $s $a (...))
+(store_field_packed $s $a (...) (...))
+(load_elem_packed_s $s $a (...))
+(load_elem_packed_u $s $a (...))
+(store_elem_packed $s $a (...) (...))
+```
+
+
+### Mutability
+
+Fields and elements can either be immutable or *mutable*:
+```
+(type $s (struct (field $a (mut i32)) (field $b i32)))
+(type $a (array (mut i32)))
+```
+Store operators are only valid when targeting a mutable field or element.
+
+Immutability is needed to enable the safe and efficient [subtyping](#subtyping), especially as needed for the [objects](#objects-and-mehtod-tables) use case.
+
+
+### Nullability
+
+By default references cannot be null,
+avoiding any runtime overhead for null checks when using them.
+
+Nullable references are available as separate types called `optref`.
+
+TODO: Design a casting operator that avoids the need for control-flow sensitive typing.
+
+
+### Defaultability
+
+Most value types, including all numeric types and nullable references are *defaultable*, which means that they have 0/null as a default value.
+Other reference types are not defaultable.
+
+Certain restrictions apply to non-defaultable types:
+
+* Local declarations of non-defaultable type must have an initializer.
+* Allocations of aggregates with non-defaultable fields or elements must have initializers.
+
+Objects whose members all have _mutable_ and _defaultable_ type may be allocated without initializers:
+```
+(type $s (struct (field $a (mut i32)) (field (mut (ref $s)))))
+(type $a (array (mut f32)))
+
+(new_default_struct $s)
+(new_default_array $a (i32.const 100))
+```
+
+
+### Sharing
+
+TODO: Distinguish types safe to share between threads in the type system.
+
+
+## Other Reference Types
+
+### Universal Type
+
+The type `anyref` can hold references of any reference type.
+It can be formed via [up casts](#casting),
+and the original type can be recovered via [down casts](#casting).
+
+
+### Foreign References
+
+A new built-in value type called `foreignref` represents opaque pointers to objects on the _embedder_'s heap.
```

Comment:
On this subject, I also think we should have the ability to have a `foreignref<T>`.  This type will allow calls to imported Web API functions to also mention `foreignref<T>` in their signature and the result will be really fast, cast-free calls from wasm into Web APIs.  The "T" would just be, I think, just an arbitrary byte string and dynamic casts would be required to go between any foreign<T> to foreign<U> where T != U, and whether the cast succeeds is up to the embedder.